herd/AArch64Sem.ml

```diff
@@ -916,7 +916,17 @@ module Make
                    add_setbits (is_zero ipte.af_v) "af:0" set_af m
                  else m
               | Dir.R ->
-                  if ha then
+                  if hd && updatedb then begin
+                    (* The case of a failed CAS with no write, but with a db update *)
+                    M.altT (
+                      add_setbits
+                        (m_op Op.Or (is_zero ipte.af_v) (is_zero ipte.db_v))
+                        "af:0 || db:0"
+                        set_afdb m
```

Comment:
Excellent to turn this into subroutines. I will check in your exact suggestion. (Not sure about embedding "ifs": the choice between subroutines depends on the comparisons, so it may look more awkward.)